### PR TITLE
fix: requestPromise to include jqXHR

### PR DESCRIPTION
**Pull request by ayesha-omarali**

### CODE REVIEW

**Review by lynnagara — APPROVED** [review]

File: src/sentry/static/sentry/app/api.jsx [hunk 1]
```diff
@@ -193,11 +193,11 @@ export class Client {
 
   requestPromise(path, options = {}) {
     return new Promise((resolve, reject) => {
+      const {isjqXHR} = options;
```
Comment (billyvg) [inline src/sentry/static/sentry/app/api.jsx]: Can we name this something more generic like `includeAllArgs` and destructure this in the function parameters? This way it doesn't get unnecessarily get passed to `this.request`:

`requestPromise(path, {includeAllArgs, ...options} = {}) {`

File: src/sentry/static/sentry/app/api.jsx [hunk 2]
```diff
@@ -193,11 +193,11 @@ export class Client {
...
       this.request(path, {
         ...options,
-        success: (data, ...args) => {
-          // This fails if we need jqXhr :(
-          resolve(data);
+        success: (data, _, jqXHR) => {
+          isjqXHR ? resolve({data, jqXHR}) : resolve(data);
```
Comment (billyvg) [inline src/sentry/static/sentry/app/api.jsx]: Can we resolve as an array, so it behaves a bit more similar to the success response in that you have positional args and not named args.

**Comment by lynnagara** [comment]
Probably need to update mocks

File: src/sentry/static/sentry/app/__mocks__/api.jsx [hunk 3]
```diff
@@ -68,14 +68,19 @@ class Client {
...
+    return new Promise((resolve, reject) => {
+      const {includeAllArgs} = options;
+      this.request(path, {
         ...options,
-        success: resolve,
-        error: reject,
-      })
-    );
+        success: (data, _, includeAllArgs) => {
+          includeAllArgs ? resolve([data, includeAllArgs]) : resolve(data);
```
Comment (lynnagara) [inline src/sentry/static/sentry/app/__mocks__/api.jsx]: Did you mean to rename the 3rd argument in the success callback also to `includeAllArgs`? This no longer refers to the correct includeAllArgs option passed into requestPromise

**Review by lynnagara — CHANGES_REQUESTED** [review]

File: src/sentry/static/sentry/app/api.jsx [hunk 4]
```diff
@@ -191,13 +191,12 @@ export class Client {
     return this.activeRequests[id];
   }
 
-  requestPromise(path, options = {}) {
+  requestPromise(path, {includeAllArgs, ...options}) {
```
Comment (lynnagara) [inline src/sentry/static/sentry/app/api.jsx]: I think you'll still need the default parameter `{}`, since the 2nd argument should be optional

**Review by lynnagara — APPROVED** [review]
💯

File: src/sentry/static/sentry/app/api.jsx [hunk 5]
```diff
@@ -191,13 +191,12 @@ export class Client {
...
 
-  requestPromise(path, options = {}) {
+  requestPromise(path, {includeAllArgs, ...options} = {}) {
     return new Promise((resolve, reject) => {
       this.request(path, {
         ...options,
-        success: (data, ...args) => {
-          // This fails if we need jqXhr :(
-          resolve(data);
+        success: (data, _, allArgs) => {
```
Comment (billyvg) [inline src/sentry/static/sentry/app/api.jsx]: So we're ignoring the middle argument here `_`, let's make sure we resolve with exactly the same args so we avoid this problem again if something eventually needs this middle arg.

File: src/sentry/static/sentry/app/api.jsx [hunk 6]
```diff
@@ -191,13 +191,12 @@ export class Client {
...
-  requestPromise(path, options = {}) {
+  requestPromise(path, {includeAllArgs, ...options} = {}) {
     return new Promise((resolve, reject) => {
       this.request(path, {
         ...options,
-        success: (data, ...args) => {
-          // This fails if we need jqXhr :(
-          resolve(data);
+        success: (data, successString, allArgs) => {
+          includeAllArgs ? resolve([data, allArgs, successString]) : resolve(data);
```
Comment (lynnagara) [inline src/sentry/static/sentry/app/api.jsx]: We should resolve in the same order
```suggestion
          includeAllArgs ? resolve([data, successString, allArgs]) : resolve(data);
```

File: src/sentry/static/sentry/app/api.jsx [hunk 7]
```diff
@@ -191,13 +191,12 @@ export class Client {
...
 
-  requestPromise(path, options = {}) {
+  requestPromise(path, {includeAllArgs, ...options} = {}) {
     return new Promise((resolve, reject) => {
       this.request(path, {
         ...options,
-        success: (data, ...args) => {
-          // This fails if we need jqXhr :(
-          resolve(data);
+        success: (data, successString, allArgs) => {
```
Comment (billyvg) [inline src/sentry/static/sentry/app/api.jsx]:
```suggestion
        success: (data, ...args) => {
```

And then you can `resolve([data, ...args])` - this way you don't need to worry about naming and order

Comment (ayesha-omarali) [reply src/sentry/static/sentry/app/api.jsx]: do we really need successString? it's pretty useless, it's a string that says "success", idk if we'd run into a problem like this again for that. 
is it just better practice to do it this way? it seems like it's unnecessary info we can just drop.

File: src/sentry/static/sentry/app/api.jsx [hunk 8]
```diff
@@ -191,13 +191,12 @@ export class Client {
...
 
-  requestPromise(path, options = {}) {
+  requestPromise(path, {includeAllArgs, ...options} = {}) {
     return new Promise((resolve, reject) => {
       this.request(path, {
         ...options,
         success: (data, ...args) => {
-          // This fails if we need jqXhr :(
-          resolve(data);
+span class="pl-smi">           includeAllArgs ? resolve([data, successString, allArgs]) : resolve(data);
```
Comment (lynnagara) [inline src/sentry/static/sentry/app/api.jsx]: Shouldn't this be `resolve([data, ...args])` to match?

Also is `span class="pl-smi">` a typo?

**Review by lynnagara — CHANGES_REQUESTED** [review]

File: src/sentry/static/sentry/app/api.jsx (hunk 8 above)
Comment (ayesha-omarali) [reply src/sentry/static/sentry/app/api.jsx]: yeah it was.

**Review by lynnagara — APPROVED** [review]
Looks good to me

**Pull request merged**